yi-core/src/Yi/CompletionTree.hs

```diff
@@ -1,59 +1,112 @@
--- | This is a little helper for completion interfaces.
+{-# LANGUAGE TupleSections #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE GeneralizedNewtypeDeriving #-}
 
-module Yi.CompletionTree (
-        CompletionTree(CT),
-        stepTree, obvious, mergeTrees, listToTree, complete)
-   where
+{- |
+Module      :  Yi.CompletionTree
+License     :  GPL-2
+Maintainer  :  yi-devel@googlegroups.com
+Stability   :  experimental
+Portability :  portable
 
-import Control.Arrow (first)
-import Data.List     (find, groupBy, intercalate, sortBy)
+Little helper for completion interfaces.
 
--- inside a completion tree, the a's must be unique on each level
+Intended to be imported qualified:
 
-data CompletionTree a = CT [(a,CompletionTree a)]
+>import qualified Yi.CompletionTree as CT
+-}
+module Yi.CompletionTree ( 
+  CompletionTree (CT), Completable (..),
+  fromCompletable, toCompletable, 
+  update, obvious, pretty
+  ) where
 
-instance (Show a) => Show (CompletionTree a) where
-  show = show'
+import           Control.Arrow   (first) 
+import           Data.Function   (on)
+import           Data.List       (partition, maximumBy, intercalate)
+import qualified Data.Map.Strict as Map
+import           Data.Map.Strict (Map)
+import           Data.Maybe      (isJust, fromJust, listToMaybe, catMaybes)
+import           Data.Monoid     ((<>))
+import           Data.Text       (Text)
 
-show' :: (Show a) => CompletionTree a -> String
-show' (CT []) = []
-show' (CT [(a,st)]) = shows a $ show' st
-show' (CT trees) = "[" ++ intercalate "|" (map (\(x,y)->shows x $ show' y) trees) ++ "]"
+-- For completable instances.
+import qualified Data.List       as L (isPrefixOf, inits, stripPrefix)
+import qualified Data.Text       as T (isPrefixOf, inits, stripPrefix)
 
-compareBy :: (Ord b) => (a->b)->a->a->Ordering
-compareBy f a b = compare (f a) (f b)
+-- | A CompletionTree is a map of parital completions.
+newtype CompletionTree a = CT (Map a (CompletionTree a)) deriving (Monoid, Show)
```

Comment:
Also it would be nice to have an `empty` and `singleton` function for this data structure and use that elsewhere. In general, we have an improvement in readability (I think) from the original but the diff is basically useless (these are practically complete reimplementations) and I'm unfamiliar with the original code, so any documentation and readability is a huge help.